### PR TITLE
Assignment4: Oscar & Leo

### PR DESCRIPTION
**Pull request by Oscar-Ruelas**
Collab w/ @LeoMarlow

### CODE REVIEW

File: Logger.Tests/StorageTests.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+using Xunit;
```
Comment (AustinH5544) [inline Logger.Tests/StorageTests.cs:1]: I know it's not officially due yet but it does say to make some tests on the storage class. It also doesn't have a check or x by it so I don't know if it is required but it is in there.

**Review by AustinH5544 — COMMENTED** [review]
Define a full name record (first, last, middle) handling optional and null appropriately. ✔
Provide a comment on the full name record on why you selected to define a value or a reference type and ✔
Provide a comment on the full name record on why or why not the type is immutable. ✔
Define an IEntity interface: ✔
Add an Id property of type Guid that is an init-only setter. ✔
Add a Name property that is string. ✔
Define an abstract base class that implements IEntity - appropriately choosing to implement the interface explicitly or implicitly. ✔
Do not implement the Name property in this abstract class. ✔
Do force any derived classes to provide an implementation for Name. ✔
Define book, student, and employee records - all with entity capabilities. ✔
Choose a thoughtful approach for the value of Name for each type of entity. In all cases, there should be no backing field for Name (not even one generated by an automatically implemented property) - consider a calculated property. ✔
Provide a comment on each interface member in each entity explaining why you implemented it implicitly or explicitly. ✔
You should consider the relationship between Student and Employee and refactor the common code shared between them. ✔
Given the properties on the full name record, you should consider which entities it makes sense to use it. ✔
Test that the equality behavior on these entities behaves as expected. ❌✔ // Kevin showed us how to do this in class today at the end(10-24)
Provide a comment on each interface method explaining why you implemented it implicitly or explicitly.
Unit test the Storage class using various entities to ensure that there are no bugs.
Code review (using TODO: comments if there are any issues - which there may not be) all logger-related classes.)
Pay special attention to the CreateLogger methods throughout.
Fundamentals
Be sure you enable:
Enable Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Ensure that you turn on Nullability (Nullable) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔
Extra Credit
Refactor your project files to put common properties and/or items into a single file. ❌✔
Convert ILogger.CreateLogger to a generic method. ❌✔
Returns the logger classes type that is to be created. ❌✔
Has a type parameter for the configuration type ❌✔

Great job on this!! Couple minor things I saw but your code looks really good. Keep up the good work ya'll!!

File: Logger.Tests/StorageTests.cs (hunk 1 above)
Comment (LeoMarlow) [reply Logger.Tests/StorageTests.cs:1]: Fixed and filled

File: Logger/Student.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+namespace Logger;
+
+
```
[the remark this answers is not in the document]
Comment (LeoMarlow) [reply Logger/Student.cs:3]: Email and school year should be Assigned by default by the school and when you change majors your School year should be a new year putting you on the new catalog year. These are initialized at the constructor.

File: Logger/FullName.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+   Using a reference type allows the object to be passed between methods or components efficiently, 
+   avoiding unnecessary memory duplication while maintaining immutability.
+*/
+
+/*
+    The FullName record is immutable because all its properties are read-only and can only be set during
+    initialization.This ensures that once a FullName object is created, its state cannot be modified.
+*/
+
+//credits: Programming C# 12: Build Cloud, Web, and Desktop Applications by Ian Griffiths
```
[the remark this answers is not in the document]
Comment (LeoMarlow) [reply Logger/FullName.cs:22]: This is actually handled in Person

File: Logger/Person.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+namespace Logger;
+
+
+
```
[the remark this answers is not in the document]
Comment (LeoMarlow) [reply Logger/Person.cs:4]: This is now fixed

**Comment by Oscar-Ruelas** [comment]
Assignment4 Is NOW COMPLETE AND I HAVE SATISFIED THE REQUIREMENT TO: Test that the equality behavior on these entities behaves as expected.

File: Logger.Tests/StorageTests.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+namespace Logger.Tests;
+    public class StorageTest
+    {
+        [Fact]
+        public void Storage_AddNewStudent_Successful()
+        {
+            // Arrange  
+            Storage storage = new();
+            IEntity student = new Student("John", "Doe", "Smith",
```
Comment (KennethWhite) [inline Logger.Tests/StorageTests.cs:11]: Nit: Generally it is not needed to store an object in a variable of the interface or base type unless working with explicit interface implementations.

File: Logger/Storage.cs [hunk 6]
```diff
@@ -1,29 +1,26 @@
...
+            return Entities.Contains(item);
+        }
+        
+        public IEntity? Get(Guid expectedGuid)
         {
-            dynamic dynamicEntity = entity;
-            return dynamicEntity.Id == expectedGuid;
-        });
+            // Using dynamic to access Id
+            return Entities.FirstOrDefault(entity => ((dynamic)entity).Id == expectedGuid);
```
Comment (KennethWhite) [inline Logger/Storage.cs:24]: If all entities are derived from BaseEntity / implement IEntity there should be no need to use the `dynamic` keyword.

**Review by KennethWhite — COMMENTED** [review]
Great job! You had fantastic test coverage and your handling of null checking was very thorough. Keep up the good work!

**Pull request closed without merging**